Add ElastiCache (Valkey) message queue template

vasigorc opened the pull request:
### Description

Adds a template demonstrating AWS ElastiCache integration for message queue functionality using Redis Streams.

## What this template shows                                                                               

- Message production to Redis Streams
- Consumer group message consumption with XREADGROUP and XAUTOCLAIM
 - Message acknowledgement and pending list management
 - Contact form to message processor workflow

### Type of Change

- [x] New Example
- [ ] Example updates (Bug fixes, new features, etc.)
- [ ] Other (changes to the codebase, but not to examples)

Simple demo flow looks like this:

https://github.com/user-attachments/assets/fdacfaca-fd49-487e-aa2a-1bb30d6d2a5b

### New Example Checklist

- [x] 🛫 `npm run new-example` was used to create the example
- [ ] 📚 The template wasn't used but I carefuly read the [Adding a new example](https://github.com/vercel/examples#adding-a-new-example) steps and implemented them in the example
- [ ] 📱 Is it responsive? Are mobile and tablets considered?

Commit:
feat: Added ElastiCache (Valkey) message queue template

Demonstrates Streams with consumer groups for message queue
functionality in Vercel Functions. Includes:

- Contact form message producer (POST /api/messages)
- Consumer group message processor with XREADGROUP and XAUTOCLAIM (GET /api/messages)
- Message acknowledgement endpoint (DELETE /api/messages)
- UI components for form submission and message processing
- valkey-glide client integration with proper error handling

acarbonetto commented on solutions/message-queue-elasticache/README.md:
````diff
@@ -0,0 +1,242 @@
+Using Docker:
+
+```bash
+docker run -d -p 6379:6379 valkey/valkey-bundle:latest
````
is valkey-bundle required for this example, as opposed to valkey (https://hub.docker.com/r/valkey/valkey/) by itself?  
If so, we should include instructions on how to install bundle modules below.

vasigorc replied:
Done

acarbonetto commented on solutions/message-queue-elasticache/README.md:
````diff
@@ -0,0 +1,242 @@
+docker run -d -p 6379:6379 valkey/valkey-bundle:latest
+```
+
+Or install Valkey directly following the [official installation guide](https://valkey.io/download/).
````
downloads page has description on installing docker.

acarbonetto commented on solutions/message-queue-elasticache/app/api/messages/route.ts:
```diff
@@ -0,0 +1,314 @@
+const STREAM_NAME = 'contact-messages'
+const CONSUMER_GROUP = 'contact-processors'
+
+let client: GlideClient | null = null
```
I prefer undefined
```suggestion
let client: GlideClient | undefined
```

acarbonetto commented on solutions/message-queue-elasticache/app/api/messages/route.ts:
```diff
@@ -0,0 +1,314 @@
+  return {
+    valid: true,
+    data: {
+      name: data.name.trim(),
```
why trim?

acarbonetto commented on solutions/message-queue-elasticache/app/api/messages/route.ts:
```diff
@@ -0,0 +1,314 @@
+     * been consumed for more than this time.
+     * See more {@link https://valkey.io/commands/xautoclaim/}
+     */
+    const claimResponse = await client.customCommand([
```
Is there no xautoclaim() available?

acarbonetto commented on solutions/message-queue-elasticache/app/api/messages/route.ts:
```diff
@@ -0,0 +1,314 @@
+     * Append entry (field, value pairs) to the specified stream: {@link https://valkey.io/commands/xadd/}
+     */
+    const streamMessageId = await client.customCommand([
+      'XADD',
```
is xadd() not available?

acarbonetto commented on solutions/message-queue-elasticache/app/api/messages/route.ts:
```diff
@@ -0,0 +1,314 @@
+     * and {@link https://valkey.io/commands/xread/}
+     */
+    const response = await client.customCommand([
+      'XREADGROUP',
```
same here?  valkey-glide should offer xreadgroup()

acarbonetto reviewed (commented):
Interesting idea, to you message processing instead of a basic API.

Commit:
fix: Updated README.md

Commit:
chore: Improved response destructiring based on PR feedback

Commit:
improvement: Replaced customCommand with specific SDK method calls

jbrinkman committed:
Update solutions/message-queue-elasticache/app/api/messages/route.ts

Co-authored-by: Andrew Carbonetto <andrew.carbonetto@improving.com>

Commit:
improvement: Bumped-up NextJS version to latest

Commit:
fix: Switch to serverExternalPackages

`serverComponentsExternalPackages` moved out of `experimental` in
Next.js 15+

Commit:
fix: Switched to use client directive

getMetadata is not compatible with client components

Commit:
fix: Updated README.md to include required information

Jonathan-Improving committed:
Change type from Mixed to Starter

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Upgrade packages

plmrry commented on solutions/aws-message-queue-elasticache/package.json:
Could you please add a `packageManager` and `engines` field, with Node version of at least `24` and a pnpm version of at least `10`?

`packageManager` is required for `turborepo` compatibility.

For example:

```
  "packageManager": "pnpm@10.29.2",
  "engines": {
    "node": "24.x",
    "pnpm": "10.29.2"
  },
```

Jonathan-Improving replied:
Done — added `"packageManager": "pnpm@10.29.2"` and `"engines": { "node": "24.x", "pnpm": "10.29.2" }`.

plmrry commented on solutions/message-queue-elasticache/package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "message-queue-elasticache",
```
Given that we have other examples in this directory that start with `aws-` could we rename this (and its parent directory) to something like `aws-message-queue-elaticache`?

Jonathan-Improving replied:
Done — renamed directory and package to `aws-message-queue-elasticache`.

plmrry commented on solutions/message-queue-elasticache/postcss.config.js:
Could you use `export default` and `postcss.config.mjs`?

Jonathan-Improving replied:
Done — replaced with `postcss.config.mjs` using `export default` and `@tailwindcss/postcss` plugin.

plmrry commented on solutions/message-queue-elasticache/tailwind.config.js:
Please remove this and use Tailwind v4 (which does not use a `tailwind.config` file)

Jonathan-Improving replied:
Done — removed `tailwind.config.js` and upgraded to Tailwind v4.

plmrry commented on solutions/message-queue-elasticache/tsconfig.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
this looks like a typo.

I would recommend deleting your `tsconfig.json` and then running `pnpm dev` — Next.js will automatically generate the correct `tsconfig.json` file. If you need different options for build or testing, create separate `tsconfig` files.

Jonathan-Improving replied:
Done — deleted `tsconfig.json`. Next.js will auto-generate it on `pnpm dev`.

plmrry commented on solutions/message-queue-elasticache/vercel.json:
please install `@vercel/config` and use `vercel.ts`

More info here:

https://www.npmjs.com/package/@vercel/config

Jonathan-Improving replied:
Done — replaced `vercel.json` with `vercel.ts` using `@vercel/config`.

plmrry commented on solutions/aws-message-queue-elasticache/turbo.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://turborepo.com/schema.json",
+  "pipeline": {
+    "build": {
```
Please add `dev` here:

```
"dev": {}
```

Jonathan-Improving replied:
Done — added `"dev": {}` to the pipeline.

plmrry reviewed (changes requested):
Overall looks good, but we need to:
- Always use typescript (or `.mjs`) in config files
- Use Node 24 or higher
- Use pnpm 10 or higher
- Use tailwindcss v4
- Update `turbo`

plmrry commented on solutions/message-queue-elasticache/package.json:
```diff
@@ -0,0 +1,30 @@
+    "eslint-config-next": "^13.4.10",
+    "postcss": "^8.4.26",
+    "tailwindcss": "^3.3.3",
+    "turbo": "^1.10.8",
```
Update `turbo`, the latest version is `2.9.9`

Jonathan-Improving replied:
Done — updated turbo to `^2.9.9`.

plmrry reviewed (changes requested):
Also update `turbo`

Jonathan-Improving committed:
Incorporate changes from review

Jonathan-Improving committed:
Rename 'message-queue-elasticache ⇒ aws-message-queue-elasticache

Jonathan-Improving commented:
I'm noticing some other issues, rectifying along with addressing the review comments above.

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
fix: address QA review findings for ElastiCache message queue template

  - Fix consumer resource leak: use stable consumer name per process
  - Replace customCommand with typed xgroupCreate API
  - Add MAXLEN to xadd to cap stream growth and prevent OOM
  - Add client health check with ping/reconnect on stale connection
  - Fix type safety: GlideString types, catch(unknown), discriminated unions
  - Extract shared StreamMessage type and buildMessageResponse helper
  - Add getErrorMessage helper for safe error coercion in UI
  - Fix misleading xautoclaim comment and empty README path
  - Enable reactStrictMode for teaching template correctness

Signed-off-by: Jonathan Neufeld <jonathan.neufeld@improving.com>

Jonathan-Improving committed:
fix: Remove BUSYGROUP warning from logs

Signed-off-by: Jonathan Neufeld <jonathan.neufeld@improving.com>

plmrry reviewed (approved):
LGTM! Thanks for the updates

atao2004 commented:
Hey @vasigorc 👋

Minor suggestion: could you add `clientName: "vercel_message_queue_client"` to the `GlideClient.createClient()` config in `app/api/messages/route.ts`? This sends a `CLIENT SETNAME` command on connection, making the connection identifiable in `CLIENT LIST` and monitoring dashboards — good practice to demonstrate in a template since users will copy this pattern.

```typescript
const client = await GlideClient.createClient({
    addresses: [{host, port}],
    requestTimeout: 5000,
    clientName: "vercel_message_queue_client",
});
```

vasigorc committed:
chore: Set Valkey clientName for connection observability

vasigorc commented:
> Hey @vasigorc 👋
> 
> Minor suggestion: could you add `clientName: "vercel_message_queue_client"` to the `GlideClient.createClient()` config in `app/api/messages/route.ts`? This sends a `CLIENT SETNAME` command on connection, making the connection identifiable in `CLIENT LIST` and monitoring dashboards — good practice to demonstrate in a template since users will copy this pattern.
> 
> ```ts
> const client = await GlideClient.createClient({
>     addresses: [{host, port}],
>     requestTimeout: 5000,
>     clientName: "vercel_message_queue_client",
> });
> ```

Hi @atao2004 — good catch, agree it's a sensible pattern to show in a template. I've opened Bit-Quill/examples#2 against this PR's branch with the change plus a small regression test asserting createClient is called with clientName: "vercel_message_queue_client". Once
@jonathanneufeld merges it, the update will roll into this PR. Thanks for the review!

Jonathan-Improving committed:
Remove security review document, not intended for repository

Jonathan-Improving committed:
Missing `tsconfig.json`

Jonathan-Improving committed:
Merge pull request #2 from vasigorc/message-queue-elasticache-clean

chore: Set Valkey clientName for connection observability

Jonathan-Improving committed:
fix: Use default Valkey localhost

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Replace all inline styles with Tailwind utility classes on the contact form and message processing pages for visual polish. Override the @vercel/examples-ui footer text from "Created by" to "Built with" via CSS pseudo-element. Add sdd/ISSUES.md tracking both reviewer-blocking items as mitigated.

Jonathan-Improving committed:
style: Apply Tailwind styling and fix footer attribution

Replace all inline styles with Tailwind 4 utility classes on the contact
form and message processing pages. Add app/globals.css with proper
Tailwind 4 entrypoint (@import "tailwindcss"). Replace @vercel/examples-ui
Layout with custom nav/footer to shorten the breadcrumb and change
"Created by" to "Built with" per reviewer feedback.

Remove stale pnpm workspace self-link from lockfile.

Jonathan-Improving committed:
Provide more descriptive error message when unable to submit a message

Jonathan-Improving committed:
Improve error and validation messages

Jonathan-Improving committed:
Include agent steering and specs

plmrry reviewed (changes requested):
1. Can we remove the `sdd` directory? Or, at least:
 - Update / remove: POLICY.md, TECH.md, and ISSUES.md, don't reference Slack.
 - Rename the directory `.agents` (What does `sdd` mean?)

2. From Vercel bot:

• Whitespace-only inputs pass validation and become empty strings after trim.
• Root layout is 'use client' with no HTML metadata (no title, no viewport).
• turbo.json uses deprecated pipeline key (should be tasks for Turbo v2).
• eslint-config-next@^13 mismatched with next@^16.

plmrry commented on solutions/aws-message-queue-elasticache/sdd/POLICY.md:
```diff
@@ -0,0 +1,49 @@
+- Open PR: [#1309](https://github.com/vercel/examples/pull/1309) on `vercel/examples`
+- Template reviewer: [plmrry](https://github.com/plmrry) (Paul Murray)
+- Post-merge: requires manual approval by v0 team to appear on [Vercel Templates homepage](https://vercel.com/templates)
+- Any code change must land in **both** the PR branch and the [demo fork](https://github.com/Jonathan-Improving/vercel-valkey-message-queue-demo) — see TECH.md Deployment section for the backport workflow
```
```suggestion

```

plmrry commented on solutions/aws-message-queue-elasticache/sdd/ISSUES.md:
```diff
@@ -0,0 +1,24 @@
+
+The demo footer displayed "Created by Vercel". Paul flagged this in Slack (Jun 29, 2026): should say "with Vercel" or "on Vercel", not "by Vercel". The text is hardcoded in the `@vercel/examples-ui` Layout component footer with no override prop.
+
+**Resolution**: Added a `<style>` block in `app/layout.tsx` that hides the span text via `visibility: hidden` and injects "Built with" via `::after` pseudo-element. The Layout component is still used (preserving the nav bar), only the footer text is overridden.
```
```suggestion

```

plmrry commented on solutions/aws-message-queue-elasticache/AGENTS.md:
```diff
@@ -0,0 +1,23 @@
+
+## Documentation
+
+Project documentation lives in the `sdd/` directory.
```
What does `sdd` mean?

Can we rename that to `.agents`?

```suggestion
Project documentation lives in the `.agents/` directory.
```

Jonathan-Improving replied:
It means "Spec-Driven Development" and contains steering documents.  I added them to bump the build and because I thought they might be useful.  They are entirely optional and can be deleted if you prefer.  I don't think it makes sense for them to be in an `.agents` sub-directory, however, perhaps `.agents/steering` though.

Jonathan-Improving replied:
✅ Done, removed the `sdd/` directory.  They're now in the branch's Git history and can be brought back if needed.

Jonathan-Improving committed:
Update solutions/aws-message-queue-elasticache/sdd/POLICY.md

Co-authored-by: Paul Murray <plmrry@gmail.com>

Jonathan-Improving committed:
Update solutions/aws-message-queue-elasticache/sdd/ISSUES.md

Co-authored-by: Paul Murray <plmrry@gmail.com>

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

Jonathan-Improving committed:
Removed Spec-Driven Development steering documents

plmrry commented on solutions/aws-message-queue-elasticache/AGENTS.md:
```diff
@@ -0,0 +1,23 @@
+
+Read all of the above, plus:
+- `sdd/POLICY.md` — Development rules, build process, and constraints you must follow
+- `sdd/TECH.md` — Technical architecture, dependencies, and module responsibilities
```
```suggestion

```

plmrry reviewed (changes requested):
Just need to remove this, or add the files back to `.agents`

Jonathan-Improving committed:
Update solutions/aws-message-queue-elasticache/AGENTS.md

Co-authored-by: Paul Murray <plmrry@gmail.com>

plmrry commented on solutions/aws-message-queue-elasticache/turbo.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "$schema": "https://turborepo.com/schema.json",
+  "pipeline": {
```
This should be "tasks," — "pipeline" is a deprecated field

Jonathan-Improving replied:
✅ Done — renamed `pipeline` to `tasks`.

plmrry commented on solutions/aws-message-queue-elasticache/app/layout.tsx:
```diff
@@ -0,0 +1,80 @@
+'use client'
```
The root layout should not use 'use client', and should either:

- provide a `<head>` with `<title>`, `<meta name="og:description /* ... */>`, etc., OR:
- export [a metadata object](https://nextjs.org/docs/app/getting-started/metadata-and-og-images#static-metadata) (this is probably the best fit here) OR:
- export [a `generateMetadata` function](https://nextjs.org/docs/app/getting-started/metadata-and-og-images#generated-metadata)

If something needs 'use client', it should go in a child of `layout.tsx`, not in the layout itself.

Jonathan-Improving replied:
✅ Done — removed `'use client'` and added static `metadata` (title + description) and `viewport` exports.

plmrry commented on solutions/aws-message-queue-elasticache/package.json:
```diff
@@ -0,0 +1,39 @@
+    "@vercel/config": "^0.3.0",
+    "autoprefixer": "^10.4.14",
+    "eslint": "^8.45.0",
+    "eslint-config-next": "^13.4.10",
```
This should match the `next` version

```suggestion
    "eslint-config-next": "^16.2.1",
```

Jonathan-Improving replied:
✅ Done — bumped to `"eslint-config-next": "^16.2.1"`.

plmrry commented on solutions/aws-message-queue-elasticache/app/api/messages/helpers.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+  if (d.message.length > 10000) {
+    return { valid: false, error: 'Message must be 10000 characters or fewer' }
+  }
```
You should check that the string length of these strings is not 0, after trimming — empty strings will pass validation:

```
{
  "name": " "
}
```

Jonathan-Improving replied:
✅ Done — validation now checks `!d.name.trim()` (and same for email/message) so whitespace-only strings are rejected before reaching the stream.

plmrry reviewed (changes requested):
Just a few more changes, thanks

Jonathan-Improving committed:
Removed AGENTS.md

Jonathan-Improving committed:
Fix review issues: whitespace validation, layout metadata, turbo.json tasks, eslint-config-next version

Jonathan-Improving commented on solutions/message-queue-elasticache/README.md:
```diff
@@ -0,0 +1,242 @@
+- Single consumer group prevents message duplication
+- Message acknowledgment ensures reliable processing
+- Refreshing the page won't cause duplicate processing
+- Demonstrates how ElastiCache Streams support reliable serverless workflows
```
✅ Done — added the 60-second auto-claim timeout to both the Key Features list and the GET endpoint description in the README.

Jonathan-Improving committed:
Add 60-second auto-claim timeout to README

plmrry reviewed (approved):
LGTM!

Jonathan-Improving committed:
Merge branch 'main' into message-queue-elasticache-clean

mykola-pereyma commented:
@plmrry can you please close this PR as duplicate https://github.com/vercel/examples/pull/1533 has been merged.

plmrry commented:
Duplicated and merged in #1533 by a vercelian (me) for security reasons.
